server/test/src/main/java/io/deephaven/server/test/FlightMessageRoundTripTest.java

```diff
@@ -387,6 +380,133 @@ public void testGetSchema() {
         }
     }
 
+    @Test
+    public void testDoExchangeSnapshot() {
+        final String staticTableName = "flightInfoTest";
+        final Table table = TableTools.emptyTable(10).update("I = i", "J = i + 0.01");
+
+        try (final SafeCloseable ignored = LivenessScopeStack.open(scriptSession, false)) {
+            // stuff table into the scope
+            scriptSession.setVariable(staticTableName, table);
+
+            // build up a snapshot request
+            byte[] magic = new byte[] {100, 112, 104, 110}; // equivalent to '0x6E687064' (ASCII "dphn")
+
+            FlightDescriptor fd = FlightDescriptor.command(magic);
+
+            try (FlightClient.ExchangeReaderWriter erw = client.doExchange(fd);
+                    final RootAllocator allocator = new RootAllocator(Integer.MAX_VALUE)) {
+
+                final FlatBufferBuilder metadata = new FlatBufferBuilder();
+
+                int optOffset =
+                        BarrageSnapshotOptions.createBarrageSnapshotOptions(metadata, ColumnConversionMode.Stringify,
+                                false, 1000);
+
+                final int ticOffset =
+                        BarrageSnapshotRequest.createTicketVector(metadata,
+                                ScopeTicketHelper.nameToBytes(staticTableName));
+                BarrageSnapshotRequest.startBarrageSnapshotRequest(metadata);
+                BarrageSnapshotRequest.addColumns(metadata, 0);
+                BarrageSnapshotRequest.addViewport(metadata, 0);
+                BarrageSnapshotRequest.addSnapshotOptions(metadata, optOffset);
+                BarrageSnapshotRequest.addTicket(metadata, ticOffset);
+                metadata.finish(BarrageSnapshotRequest.endBarrageSnapshotRequest(metadata));
+
+                final FlatBufferBuilder wrapper = new FlatBufferBuilder();
+                final int innerOffset = wrapper.createByteVector(metadata.dataBuffer());
+                wrapper.finish(BarrageMessageWrapper.createBarrageMessageWrapper(
+                        wrapper,
+                        0x6E687064, // the numerical representation of the ASCII "dphn".
+                        BarrageMessageType.BarrageSnapshotRequest,
+                        innerOffset));
+
+                // extract the bytes and package them in an ArrowBuf for transmission
+                byte[] msg = wrapper.sizedByteArray();
+                ArrowBuf data = allocator.buffer(msg.length);
+                data.writeBytes(msg);
+
+                erw.getWriter().putMetadata(data);
+                erw.getWriter().completed();
+
+                // read everything from the server
+                while (erw.getReader().next()) {
+                    // NOP
```

Comment:
I suggest counting the messages. I see below that you expect the root to match the expected row count. I think a given root is applicable immediately after `next()` returns. I'm kind of surprised it is not nulled out after `next()` returns `false`. IMO, I would label that a flight bug (maybe it doesn't matter if they are using their own allocator). Either way, it probably makes sense to error if the number of messages is unexpected before erroring if the number of rows is unexpected. (Fragmentation, for example, is likely to break big snapshots up.)